src/cryptonote_core/blockchain.cpp

```diff
@@ -3312,118 +3313,201 @@ void Blockchain::flush_invalid_blocks() {
-
-            // NOTE: Determine if the node was unlocked/deregistered or only exists in the contract
-            const eth::bls_public_key& it = result[index];
-            bool protocol_dereg = false;
-            std::optional<uint64_t> unlock_or_dereg_height = {};
-            uint32_t ip = 0;
-            uint16_t port = 0;
-            service_node_list.for_each_recently_removed_node([&](const auto& node) {
-                if (it != node.info.bls_public_key)
                     return false;
-                protocol_dereg =
-                        node.type ==
-                        service_nodes::service_node_list::recently_removed_node::type_t::deregister;
-                unlock_or_dereg_height = node.height;
-                ip = node.public_ip;
-                port = node.qnet_port;
-                return true;
+                } else {
+                    static_assert(
+                            std::is_same_v<Event, eth::event::ServiceNodeExitRequest> ||
+                            std::is_same_v<Event, eth::event::ServiceNodeExit> ||
+                            std::is_same_v<Event, eth::event::StakingRequirementUpdated> ||
+                            std::is_same_v<Event, eth::event::ServiceNodePurge>);
+                }
             });
 
-            // NOTE: Generate a reason string
-            std::string removable_reason;
-            if (unlock_or_dereg_height) {
-                removable_reason = "node {}; height {} {}:{}"_format(
-                        protocol_dereg ? "dereg" : "expired",
-                        *unlock_or_dereg_height,
-                        epee::string_tools::get_ip_string_from_int32(ip),
-                        port);
-            } else {
-                removable_reason = "node exists only in contract";
-            }
+    // Incoming, unconfirmed registrations are not removable:
+    if (pending_reg) {
+        log::debug(
+                logcat,
+                "{} is not removable: found a pending but unconfirmed new service node event with "
+                "that BLS pubkey",
+                bls_pubkey);
+        return false;
+    }
 
-            fmt::format_to(
-                    std::back_inserter(buffer), " {:04d} {} ({})", index, it, removable_reason);
-        }
+    // Case 2: not in the registered SN list, recently removed, or incoming unconfirmed, but *is* in
+    // the contract: this is both removable and liquidatable.
+    if (service_node_list.find_public_key_registered(bls_pubkey)) {
+        log::debug(
+                logcat,
+                "{} is not removable: pubkey belongs to a registered service node",
+                bls_pubkey);
+        return false;
+    }
+
+    if (!m_l2_tracker)
+        log::debug(
+                logcat,
+                "No L2 tracker configured; skipping in-contract-but-not-oxend removable check of "
+                "{}",
+                bls_pubkey);
```

Comment:
```suggestion
    if (!m_l2_tracker) {
        log::debug(
                logcat,
                "No L2 tracker configured; skipping in-contract-but-not-oxend removable check of "
                "{}",
                bls_pubkey);
        return false;
    }
```

Missing early return